worker/uniter/relation/relations_test.go

```diff
@@ -946,11 +948,207 @@ func (s *relationsSuite) TestSubSubOtherRelationDyingNotDestroyed(c *gc.C) {
 		},
 	}
 
-	rr := relation.NewRelationsResolver(r)
+	_, rr := relation.NewRelationResolvers(r, nil)
 	_, err = rr.NextOp(localState, remoteState, &mockOperations{})
 	c.Assert(err, jc.ErrorIsNil)
 
 	// Check that we didn't try to make a destroy call (the apiCaller
 	// should panic in that case anyway).
 	assertNumCalls(c, &numCalls, expectedCalls)
 }
+
+func principalWithSubordinateAPICalls() []apiCall {
+	relationStatusResults := params.RelationUnitStatusResults{Results: []params.RelationUnitStatusResult{{
+		RelationResults: []params.RelationUnitStatus{{
+			RelationTag: "relation-wordpress:juju-info nrpe:general-info",
+			InScope:     true,
+		},
+		}}}}
+	relationUnits1 := params.RelationUnits{RelationUnits: []params.RelationUnit{
+		{Relation: "relation-wordpress.juju-info#nrpe.general-info", Unit: "unit-nrpe-0"},
+	}}
+	relationResults1 := params.RelationResults{
+		Results: []params.RelationResult{{
+			Id:               1,
+			Key:              "wordpress:juju-info nrpe:general-info",
+			Life:             life.Alive,
+			OtherApplication: "wordpress",
+			Endpoint: params.Endpoint{
+				ApplicationName: "nrpe",
+				Relation: params.CharmRelation{
+					Name:      "general-info",
+					Role:      string(charm.RoleRequirer),
+					Interface: "juju-info",
+					Scope:     "container",
+				},
+			},
+		}},
+	}
+	relationStatus1 := params.RelationStatusArgs{Args: []params.RelationStatusArg{{
+		UnitTag:    "unit-nrpe-0",
+		RelationId: 1,
+		Status:     params.Joined,
+	}}}
+
+	return []apiCall{
+		uniterAPICall("Refresh", nrpeUnitEntity, params.UnitRefreshResults{Results: []params.UnitRefreshResult{{Life: life.Alive, Resolved: params.ResolvedNone}}}, nil),
+		uniterAPICall("GetPrincipal", nrpeUnitEntity, params.StringBoolResults{Results: []params.StringBoolResult{{Result: "unit-wordpress-0", Ok: true}}}, nil),
+		uniterAPICall("RelationsStatus", nrpeUnitEntity, relationStatusResults, nil),
+		uniterAPICall("Relation", relationUnits1, relationResults1, nil),
+		uniterAPICall("Relation", relationUnits1, relationResults1, nil),
+		uniterAPICall("Watch", nrpeUnitEntity, params.NotifyWatchResults{Results: []params.NotifyWatchResult{{NotifyWatcherId: "1"}}}, nil),
+		uniterAPICall("EnterScope", relationUnits1, noErrorResult, nil),
+		uniterAPICall("SetRelationStatus", relationStatus1, noErrorResult, nil),
+	}
+}
+
+func (s *relationsSuite) TestPrincipalDyingDestroysSubordinates(c *gc.C) {
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+
+	var numCalls int32
+	apiCalls := principalWithSubordinateAPICalls()
```

Comment:
Yeah... my head hurts trying to read these tests. I only added this one because I realized while shuffling parts of the code around that this pre-existing codepath was not exercised by tests (a.k.a "wait, shouldn't some of these tests fail now that I removed this bit?")

My view is that all of these tests need to be rewritten to use mocks but that would require too much effort and is probably outside of the scope for this PR.